packages/realm-server/prerender/prerenderer.ts

```diff
@@ -0,0 +1,480 @@
+import {
+  type PrerenderMeta,
+  type CardErrorJSONAPI,
+  uuidv4,
+  logger,
+  Deferred,
+} from '@cardstack/runtime-common';
+import puppeteer, {
+  type Browser,
+  type BrowserContext,
+  type Page,
+} from 'puppeteer';
+import { createJWT } from '../jwt';
+import {
+  captureResult,
+  isRenderError,
+  renderAncestors,
+  renderHTML,
+  renderIcon,
+  renderMeta,
+  RenderCapture,
+  RenderError,
+  withTimeout,
+  transitionTo,
+} from './utils';
+
+const log = logger('prerenderer');
+const boxelHostURL = process.env.BOXEL_HOST_URL ?? 'http://localhost:4200';
+
+export type PermissionsMap = {
+  [realm: string]: ('read' | 'write' | 'realm-owner')[];
+};
+
+export interface RenderResponse extends PrerenderMeta {
+  isolatedHTML: string | null;
+  atomHTML: string | null;
+  embeddedHTML: Record<string, string> | null;
+  fittedHTML: Record<string, string> | null;
+  iconHTML: string | null;
+  error?: RenderError;
+}
+
+export class Prerenderer {
+  #browser: Browser | null = null;
+  #pool = new Map<
+    string,
+    {
+      realm: string;
+      context: BrowserContext;
+      page: Page;
+      pageId: string;
+      lastUsedAt: number;
+    }
+  >();
+  #lru = new Set<string>();
+  #pendingByRealm = new Map<string, Promise<void>>();
+  #maxPages: number;
+  #secretSeed: string;
+  #stopped = false;
+  #evictionMetrics = {
+    byRealm: new Map<string, { unusable: number; timeout: number }>(),
+  };
```

Comment:
ultimately i'd like to expose the eviction metrics to grafana so that we can more easily spot issues that are causing ember to crash during prerendering